src/animation/customGraphicKeyframeAnimation.ts

```diff
@@ -0,0 +1,161 @@
+
+// Helpers for creating keyframe based animations in custom series and graphic components.
+
+type AnimationKeyframe<T extends Record<string, any>> = T & {
+    easing?: AnimationEasing
+    percent?: number    // 0 - 1
+};
+
+type StateToRestore = Dictionary<any>;
+const getStateToRestore = makeInner<StateToRestore, Element>();
+
+export interface ElementKeyframeAnimationOption<Props extends Record<string, any>> extends AnimationOption {
+    // Animation configuration for keyframe based animation.
+    loop?: boolean
+    keyframes?: AnimationKeyframe<Props>[]
+}
+
+/**
+ * Stopped previous keyframe animation and restore the attributes.
+ * Avoid new keyframe animation starts with wrong internal state when the percent: 0 is not set.
+ */
+export function stopPreviousKeyframeAnimationAndRestore(el: Element) {
+    // Stop previous keyframe animation.
+    el.stopAnimation('keyframe');
+    // Restore
+    el.attr(getStateToRestore(el));
+}
+
+export function applyKeyframeAnimation<T extends Record<string, any>>(
+    el: Element,
+    animationOpts: ElementKeyframeAnimationOption<T> | ElementKeyframeAnimationOption<T>[],
+    animatableModel: Model<AnimationOptionMixin>
+) {
+    if (!animatableModel.isAnimationEnabled() || !animationOpts) {
+        return;
+    }
+
+    if (isArray(animationOpts)) {
+        each(animationOpts, singleAnimationOpts => {
+            applyKeyframeAnimation(el, singleAnimationOpts, animatableModel);
+        });
+        return;
+    }
+
+    const keyframes = animationOpts.keyframes;
+    let duration = animationOpts.duration;
+
+    if (animatableModel && duration == null) {
+        const config = getAnimationConfig('enter', animatableModel, 0);
+        duration = config && config.duration;
+    }
+
+    if (!keyframes || !duration) {
+        return;
+    }
+
+    const stateToRestore: StateToRestore = getStateToRestore(el);
+
+    each(ELEMENT_ANIMATABLE_PROPS, (targetPropName) => {
+        if (targetPropName && !(el as any)[targetPropName]) {
+            return;
+        }
+
+        let animator: ReturnType<Element['animate']>;
+        let endFrameIsSet = false;
+
+        // Sort keyframes by percent.
+        keyframes.sort((a, b) => a.percent - b.percent);
+
+        each(keyframes, kf => {
+            // Stop current animation.
+            const animators = el.animators;
+            const kfValues = targetPropName ? kf[targetPropName] : kf;
+
+            if (__DEV__) {
+                if (kf.percent >= 1) {
+                    endFrameIsSet = true;
+                }
+            }
+
+            if (!kfValues) {
+                return;
+            }
+
+            let propKeys = keys(kfValues);
+            if (!targetPropName) {
+                // PENDING performance?
+                propKeys = filter(
+                    propKeys, key => key !== 'percent' && key !== 'easing'
+                    && key !== 'shape' && key !== 'style' && key !== 'extra'
+                );
```

Comment:
It's better to use `include`?
```ts
!['percent', 'easing', 'shape', 'style', 'extra'].include(key)
```

Reply:
Thanks for the suggestion. Changed to using `indexOf` to check if key is included. `includes` is not supported in IE. Which we haven't dropped officialy yet.